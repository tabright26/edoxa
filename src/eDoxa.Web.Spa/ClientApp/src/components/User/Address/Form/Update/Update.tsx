import React, { FunctionComponent } from "react";
import { FormGroup, Col, Form } from "reactstrap";
import { Field, reduxForm, InjectedFormProps, FormErrors } from "redux-form";
import Input from "components/Shared/Input";
import Button from "components/Shared/Button";
import FormField from "components/Shared/Form/Field";
import { UPDATE_USER_ADDRESS_FORM } from "utils/form/constants";
import { compose } from "recompose";
import FormValidation from "components/Shared/Form/Validation";
import { updateUserAddress } from "store/actions/identity";
import { throwSubmissionError } from "utils/form/types";
import {
  LINE1_REQUIRED,
  line1Regex,
  LINE1_INVALID,
  line2Regex,
  LINE2_INVALID,
  CITY_REQUIRED,
  cityRegex,
  CITY_INVALID,
  stateRegex,
  STATE_INVALID,
  postalRegex,
  POSTAL_INVALID
} from "validation";
import { connect, MapStateToProps } from "react-redux";
import { AddressId } from "types";
import { RootState } from "store/types";
import { AxiosActionCreatorMeta } from "utils/axios/types";

interface StateProps {}

interface FormData {
  line1: string;
  line2: string;
  city: string;
  state: string;
  postalCode: string;
}

interface OutterProps {
  addressId: AddressId;
  handleCancel: () => void;
}

type InnerProps = InjectedFormProps<FormData, Props> & StateProps;

type Props = InnerProps & OutterProps;

const CustomForm: FunctionComponent<Props> = ({
  handleSubmit,
  error,
  handleCancel
}) => (
  <Form onSubmit={handleSubmit}>
    {error && <FormValidation error={error} />}
    <FormGroup>
      <FormField.Country disabled={true} />
    </FormGroup>
    <Field
      type="text"
      name="line1"
      label="Address line 1"
      formGroup={FormGroup}
      component={Input.Text}
    />
    <Field
      type="text"
      name="line2"
      label="Address line 2 (optional)"
      formGroup={FormGroup}
      component={Input.Text}
    />
    <Field
      type="text"
      name="city"
      label="City"
      formGroup={FormGroup}
      component={Input.Text}
    />
    <FormGroup row className="my-0">
      <Col xs="8">
        <Field
          type="text"
          name="state"
          label="State"
          formGroup={FormGroup}
          component={Input.Text}
        />
      </Col>
      <Col xs="4">
        <Field
          type="text"
          name="postalCode"
          label="Postal Code"
          formGroup={FormGroup}
          component={Input.Text}
        />
      </Col>
    </FormGroup>
    <FormGroup className="mb-0">
      <Button.Save className="mr-2" />
      <Button.Cancel onClick={() => handleCancel()} />
    </FormGroup>
  </Form>
);

const mapStateToProps: MapStateToProps<StateProps, Props, RootState> = (
  state,
  ownProps
) => {
  const { data } = state.root.user.addressBook;
  const address = data.find(address => address.id === ownProps.addressId);
  return {
    initialValues: address
  };
};

const enhance = compose<InnerProps, OutterProps>(
  connect(mapStateToProps),
  reduxForm<FormData, Props>({
    form: UPDATE_USER_ADDRESS_FORM,
    onSubmit: async (values, dispatch, { addressId }) => {
      try {
        return await new Promise((resolve, reject) => {
          const meta: AxiosActionCreatorMeta = { resolve, reject };
          dispatch(updateUserAddress(addressId, values, meta));
        });
      } catch (error) {
        throwSubmissionError(error);
      }
    },
    onSubmitSuccess: (result, dispatch, { handleCancel }) => handleCancel(),
    validate: values => {
      const errors: FormErrors<FormData> = {};
      if (!values.line1) {
        errors.line1 = LINE1_REQUIRED;
      } else if (!line1Regex.test(values.line1)) {
        errors.line1 = LINE1_INVALID;
      }
      if (values.line2 && !line2Regex.test(values.line2)) {
        errors.line2 = LINE2_INVALID;
      }
      if (!values.city) {
        errors.city = CITY_REQUIRED;
      } else if (!cityRegex.test(values.city)) {
        errors.city = CITY_INVALID;
      }
      if (values.state && !stateRegex.test(values.state)) {
        errors.state = STATE_INVALID;
      }
      if (values.postalCode && !postalRegex.test(values.postalCode)) {
        errors.postalCode = POSTAL_INVALID;
      }
      return errors;
    }
  })
);

export default enhance(CustomForm);
