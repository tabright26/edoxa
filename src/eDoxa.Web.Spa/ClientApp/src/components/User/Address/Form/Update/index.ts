import UpdateUserAddressForm from "./Update";

export default UpdateUserAddressForm;
