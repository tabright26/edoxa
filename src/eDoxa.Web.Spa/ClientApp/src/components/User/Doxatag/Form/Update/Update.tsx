import React, { FunctionComponent } from "react";
import { FormGroup, Form } from "reactstrap";
import { reduxForm, Field, InjectedFormProps, FormErrors } from "redux-form";
import Button from "components/Shared/Button";
import Input from "components/Shared/Input";
import { UPDATE_USER_DOXATAG_FORM } from "utils/form/constants";
import { compose } from "recompose";
import { ValidationSummary } from "components/Shared/ValidationSummary";
import { changeUserDoxatag } from "store/actions/identity";
import { throwSubmissionError } from "utils/form/types";
import {
  DOXATAG_REGEXP,
  DOXATAG_MIN_LENGTH,
  DOXATAG_MAX_LENGTH,
  DOXATAG_REQUIRED,
  DOXATAG_MIN_LENGTH_INVALID,
  DOXATAG_MAX_LENGTH_INVALID,
  DOXATAG_INVALID
} from "utils/form/validators";
import { connect, MapStateToProps } from "react-redux";
import { RootState } from "store/types";
import { AxiosActionCreatorMeta } from "utils/axios/types";
import produce, { Draft } from "immer";
import { UserDoxatag } from "types";
import authorizeService from "utils/oidc/AuthorizeService";

interface FormData {
  name: string;
}

interface StateProps {}

interface OutterProps {
  handleCancel: () => void;
}

type InnerProps = InjectedFormProps<FormData, Props> & StateProps;

type Props = InnerProps & OutterProps;

const CustomForm: FunctionComponent<Props> = ({
  error,
  handleSubmit,
  handleCancel
}) => (
  <Form onSubmit={handleSubmit}>
    <ValidationSummary error={error} />
    <Field
      type="text"
      name="name"
      placeholder="Name"
      formGroup={FormGroup}
      component={Input.Text}
    />
    <FormGroup className="mb-0">
      <Button.Save className="mr-2" />
      <Button.Cancel onClick={() => handleCancel()} />
    </FormGroup>
  </Form>
);

const mapStateToProps: MapStateToProps<
  StateProps,
  Props,
  RootState
> = state => {
  const { data } = state.root.user.doxatagHistory;
  const doxatags = produce(data, (draft: Draft<UserDoxatag[]>) => {
    draft.sort((left: UserDoxatag, right: UserDoxatag) =>
      left.timestamp < right.timestamp ? 1 : -1
    );
  });
  return {
    initialValues: doxatags[0] || null
  };
};

const enhance = compose<InnerProps, OutterProps>(
  connect(mapStateToProps),
  reduxForm<FormData, Props>({
    form: UPDATE_USER_DOXATAG_FORM,
    onSubmit: async (values, dispatch) => {
      try {
        return await new Promise(async (resolve, reject) => {
          const meta: AxiosActionCreatorMeta = { resolve, reject };
          dispatch(changeUserDoxatag(values, meta));
        });
      } catch (error) {
        throwSubmissionError(error);
      }
    },
    onSubmitSuccess: (result, dispatch, { handleCancel }) => {
      handleCancel();
      authorizeService.signIn({
        returnUrl: window.location.pathname
      });
    },
    validate: values => {
      const errors: FormErrors<FormData> = {};
      if (!values.name) {
        errors.name = DOXATAG_REQUIRED;
      } else if (values.name.length < DOXATAG_MIN_LENGTH) {
        errors.name = DOXATAG_MIN_LENGTH_INVALID;
      } else if (values.name.length > DOXATAG_MAX_LENGTH) {
        errors.name = DOXATAG_MAX_LENGTH_INVALID;
      } else if (!DOXATAG_REGEXP.test(values.name)) {
        errors.name = DOXATAG_INVALID;
      }
      return errors;
    }
  })
);

export default enhance(CustomForm);
