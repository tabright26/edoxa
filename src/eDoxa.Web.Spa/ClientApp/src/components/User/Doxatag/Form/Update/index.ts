import UpdateUserDoxatagForm from "./Update";

export default UpdateUserDoxatagForm;
