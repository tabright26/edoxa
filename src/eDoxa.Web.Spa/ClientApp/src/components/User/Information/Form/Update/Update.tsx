import React, { FunctionComponent } from "react";
import { Form, FormGroup } from "reactstrap";
import { Field, reduxForm, FormSection, InjectedFormProps } from "redux-form";
import Input from "components/Shared/Input";
import Button from "components/Shared/Button";
import { UPDATE_USER_INFORMATIONS_FORM } from "forms";
import { compose } from "recompose";
import FormField from "components/Shared/Form/Field";
import FormValidation from "components/Shared/Form/Validation";
import { updateUserInformations } from "store/actions/identity";
import { throwSubmissionError } from "utils/form/types";
import { connect, MapStateToProps } from "react-redux";
import { RootState } from "store/types";
import {
  PERSONALINFO_FIRSTNAME_REQUIRED,
  PERSONALINFO_FIRSTNAME_INVALID,
  personalInfoNameRegex
} from "validation";
import { AxiosActionCreatorMeta } from "utils/axios/types";

interface Props {}

interface FormData {}

interface StateProps {}

const validate = values => {
  const errors: any = {};
  if (!values.firstName) {
    errors.firstName = PERSONALINFO_FIRSTNAME_REQUIRED;
  } else if (!personalInfoNameRegex.test(values.firstName)) {
    errors.firstName = PERSONALINFO_FIRSTNAME_INVALID;
  }
  return errors;
};

async function submit(values, dispatch) {
  try {
    return await new Promise((resolve, reject) => {
      const meta: AxiosActionCreatorMeta = { resolve, reject };
      dispatch(updateUserInformations(values, meta));
    });
  } catch (error) {
    throwSubmissionError(error);
  }
}

const UpdateUserInformationsForm: FunctionComponent<InjectedFormProps<
  FormData
> &
  Props &
  any> = ({ handleSubmit, handleCancel, dispatch, error }) => (
  <Form
    onSubmit={handleSubmit(data =>
      submit(data, dispatch).then(() => handleCancel())
    )}
  >
    {error && <FormValidation error={error} />}
    <dl className="row mb-0">
      <dd className="col-sm-3 text-muted mb-0">Name</dd>
      <dd className="col-sm-9 mb-0">
        <dl className="row">
          <dt className="col-sm-4 mb-0">
            <Field name="firstName" label="First Name" component={Input.Text} />
          </dt>
          <dd className="col-sm-4 mb-0">
            <Field
              name="lastName"
              label="Last Name"
              component={Input.Text}
              disabled
            />
          </dd>
        </dl>
      </dd>
      <dd className="col-sm-3 text-muted mb-0">Date of birth</dd>
      <dd className="col-sm-9 mb-0">
        <FormGroup>
          <FormSection name="dob">
            <FormField.Year className="d-inline" width="75px" disabled />
            <span className="d-inline mx-2">/</span>
            <FormField.Month className="d-inline" width="60px" disabled />
            <span className="d-inline mx-2">/</span>
            <FormField.Day className="d-inline" width="60px" disabled />
          </FormSection>
        </FormGroup>
      </dd>
      <dd className="col-sm-3 mb-0 text-muted">Gender</dd>
      <dd className="col-sm-3 mb-0">
        <Field name="gender" label="Gender" component={Input.Text} disabled />
      </dd>
      <dd className="col-sm-9 mb-0">
        <Button.Save className="mt-3 mr-2" />
        <Button.Cancel className="mt-3" onClick={handleCancel} />
      </dd>
    </dl>
  </Form>
);

const mapStateToProps: MapStateToProps<StateProps, Props, RootState> = (
  state: RootState
) => {
  const { data } = state.root.user.information;
  return {
    initialValues: data
  };
};

const enhance = compose<any, any>(
  reduxForm<FormData, Props>({
    form: UPDATE_USER_INFORMATIONS_FORM,
    validate
  }),
  connect(mapStateToProps)
);

export default enhance(UpdateUserInformationsForm);
