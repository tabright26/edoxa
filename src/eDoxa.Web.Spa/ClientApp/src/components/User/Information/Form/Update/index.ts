import UpdateUserInformationsForm from "./Update";

export default UpdateUserInformationsForm;
