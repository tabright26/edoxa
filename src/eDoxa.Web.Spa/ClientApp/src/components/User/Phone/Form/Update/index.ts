import UpdateUserPhoneForm from "./Update";

export default UpdateUserPhoneForm;
