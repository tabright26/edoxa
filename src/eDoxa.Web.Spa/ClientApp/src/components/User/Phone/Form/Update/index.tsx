import React, { FunctionComponent } from "react";
import { FormGroup, Form } from "reactstrap";
import { Field, reduxForm, InjectedFormProps, FormErrors } from "redux-form";
import Input from "components/Shared/Input";
import Button from "components/Shared/Button";
import { UPDATE_USER_PHONE_FORM } from "utils/form/constants";
import { compose } from "recompose";
import { ValidationSummary } from "components/Shared/ValidationSummary";
import { updateUserPhone } from "store/actions/identity";
import { throwSubmissionError } from "utils/form/types";
import { connect, MapStateToProps } from "react-redux";
import { RootState } from "store/types";
import { AxiosActionCreatorMeta } from "utils/axios/types";
import {
  PHONE_REQUIRED,
  PHONE_INVALID,
  PHONE_REGEXP
} from "utils/form/validators";

interface StateProps {}

interface FormData {
  number: number;
}
interface OutterProps {
  handleCancel: () => void;
}

type InnerProps = InjectedFormProps<FormData, Props> & StateProps;

type Props = InnerProps & OutterProps;

const CustomForm: FunctionComponent<Props> = ({
  handleSubmit,
  error,
  handleCancel,
  submitting
}) => (
  <Form onSubmit={handleSubmit}>
    <ValidationSummary error={error} />
    <Field
      type="text"
      name="number"
      placeholder="Phone Number"
      formGroup={FormGroup}
      component={Input.Text}
    />
    <FormGroup className="mb-0">
      <Button.Submit loading={submitting} className="mr-2" size="sm">
        Save
      </Button.Submit>
      <Button.Cancel onClick={handleCancel} />
    </FormGroup>
  </Form>
);

const mapStateToProps: MapStateToProps<
  StateProps,
  Props,
  RootState
> = state => {
  const { data } = state.root.user.phone;
  return {
    initialValues: data
  };
};

const enhance = compose<InnerProps, OutterProps>(
  connect(mapStateToProps),
  reduxForm<FormData, Props>({
    form: UPDATE_USER_PHONE_FORM,
    onSubmit: async (values, dispatch) => {
      try {
        return await new Promise((resolve, reject) => {
          const meta: AxiosActionCreatorMeta = { resolve, reject };
          dispatch(updateUserPhone(values, meta));
        });
      } catch (error) {
        throwSubmissionError(error);
      }
    },
    onSubmitSuccess: (result, dispatch, { handleCancel }) => handleCancel(),
    validate: values => {
      const errors: FormErrors<FormData> = {};
      if (!values.number) {
        errors.number = PHONE_REQUIRED;
      } else if (!PHONE_REGEXP.test(values.number.toString())) {
        errors.number = PHONE_INVALID;
      }
      return errors;
    }
  })
);

export default enhance(CustomForm);
