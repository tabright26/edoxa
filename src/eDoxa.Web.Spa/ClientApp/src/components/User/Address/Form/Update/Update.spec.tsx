import React from "react";
import { Provider } from "react-redux";
import { ReactWrapper } from "enzyme";
import Update from "./Update";
import { configureStore } from "store";
import { FormGroup } from "reactstrap";
import Input from "components/Shared/Input";
import {
  ADDRESS_LINE1_REQUIRED,
  ADDRESS_LINE1_INVALID,
  ADDRESS_LINE2_INVALID,
  ADDRESS_CITY_REQUIRED,
  ADDRESS_CITY_INVALID,
  ADDRESS_STATE_INVALID,
  ADDRESS_POSTAL_CODE_INVALID
} from "validation";

const shallow = global["shallow"];
const mount = global["mount"];

const store = configureStore();

const createWrapper = (): ReactWrapper | any => {
  return mount(
    <Provider store={store}>
      <Update addressId="addressId" handleCancel={() => {}} />
    </Provider>
  );
};

describe("<UserAddressUpdateForm />", () => {
  it("should match the snapshot", () => {
    const shallowWrapper = shallow(
      <Provider store={store}>
        <Update addressId="addressId" handleCancel={() => {}} />
      </Provider>
    );
    expect(shallowWrapper).toMatchSnapshot();
  });

  describe("defines address update form fields", () => {
    it("renders country field", () => {
      const wrapper = createWrapper();
      const field = wrapper.findFieldByName("country");

      expect(field.prop("type")).toBe("select");
      expect(field.prop("component")).toBe(Input.Select);
    });

    test.each([
      ["line1", "Address line 1"],
      ["line2", "Address line 2 (optional)"],
      ["city", "City"],
      ["state", "State"],
      ["postalCode", "Postal Code"]
    ])("renders name fields", (name: string, label: string) => {
      const wrapper = createWrapper();
      const field = wrapper.findFieldByName(name);

      expect(field.prop("type")).toBe("text");
      expect(field.prop("label")).toBe(label);
      expect(field.prop("formGroup")).toBe(FormGroup);
      expect(field.prop("component")).toBe(Input.Text);
    });

    it("renders save button", () => {
      const wrapper = createWrapper();
      const saveButton = wrapper.findSaveButton();

      expect(saveButton.prop("type")).toBe("submit");
      expect(saveButton.text()).toBe("Save");
    });

    it("renders cancel button", () => {
      const wrapper = createWrapper();
      const cancelButton = wrapper.findCancelButton();

      expect(cancelButton.prop("type")).toBe("button");
      expect(cancelButton.text()).toBe("Cancel");
    });
  });

  describe("form validation", () => {
    describe("text fields validation", () => {
      test.each([
        ["line1", ADDRESS_LINE1_REQUIRED],
        ["city", ADDRESS_CITY_REQUIRED]
      ])("name fields blank validation", (name: string, message: string) => {
        const wrapper = createWrapper();
        const input = wrapper.findInputByName(name);
        input.simulate("blur");

        const errorPresent = wrapper.findFormFeedback(message);
        expect(errorPresent).toBeTruthy();
      });

      test.each([
        ["line1", ADDRESS_LINE1_INVALID],
        ["line2", ADDRESS_LINE2_INVALID],
        ["city", ADDRESS_CITY_INVALID],
        ["state", ADDRESS_STATE_INVALID],
        ["postalCode", ADDRESS_POSTAL_CODE_INVALID]
      ])("name fields blank validation", (name: string, message: string) => {
        const wrapper = createWrapper();
        const input = wrapper.findInputByName(name);
        input.simulate("change", { target: { value: "_!" } });

        const errorPresent = wrapper.findFormFeedback(message);
        expect(errorPresent).toBeTruthy();
      });
    });
  });
});
