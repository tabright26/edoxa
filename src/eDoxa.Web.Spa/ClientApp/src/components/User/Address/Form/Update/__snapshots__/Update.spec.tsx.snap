// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAddressUpdateForm /> should match the snapshot 1`] = `ShallowWrapper {}`;
